File: taurus-protect-sdk-typescript/tests/unit/mappers/rule-cell-codec.test.ts
/**
 * Branch/edge coverage for the cell codec that the golden vectors exercise only
 * transitively: error paths, the RawCell fallback, the magnitude encoding, and
 * cellFamily.
 */
import type { RuleCell } from '../../../src/models/rule-cell';
import { cellFamily, ruleCellFromBytes, ruleCellToBytes } from '../../../src/mappers/rule-cell-codec';
import {
  Blockchain,
  RuleDestination,
  RuleDestination_RuleDestinationType,
  RuleDestinationContractAddress,
  RuleFiatAmount,
  RuleFiatAmount_RuleFiatAmountType,
  RuleIntegerGreater,
  RuleIntegerGreater_RuleIntegerGreaterType,
  RuleUIntegerGreater,
  RuleUIntegerGreater_RuleUIntegerGreaterType,
} from '../../../src/internal/proto/request_reply';

const bytes = (u: Uint8Array): number[] => Array.from(u);

// --- Error branches ---

describe('rule cell codec error branches', () => {
  it('throws when a cell is placed in the wrong column family', () => {
    expect(() => ruleCellToBytes('RuleSource', { kind: 'FiatAmountAny' })).toThrow(/not valid for column type/);
  });

  it('throws for a negative unsigned integer', () => {
    expect(() => ruleCellToBytes('RuleUIntegerGreater', { kind: 'UIntegerGreaterValue', value: -1n })).toThrow(/non-negative/);
    expect(() => ruleCellToBytes('RuleUIntegerGreater', { kind: 'UIntegerGreaterIsEqual', value: -1n })).toThrow(/non-negative/);
  });
});

// --- RawCell fallback branches ---

describe('rule cell codec RawCell fallback', () => {
  it('preserves an unknown column type verbatim (empty and non-empty)', () => {
    const nonEmpty = ruleCellFromBytes('RuleUnknownColumn', new Uint8Array([0x08, 0x01]));
    expect(nonEmpty.kind).toBe('RawCell');
    expect(bytes(ruleCellToBytes('RuleUnknownColumn', nonEmpty))).toEqual([0x08, 0x01]);

    const empty = ruleCellFromBytes('RuleUnknownColumn', new Uint8Array());
    expect(empty.kind).toBe('RawCell');
    expect(bytes(ruleCellToBytes('RuleUnknownColumn', empty))).toEqual([]);
  });

  it('preserves a wrapper with an unknown sub-field verbatim (lossless guard)', () => {
    const data = new Uint8Array([0xc0, 0x0c, 0x2a]); // field 200, varint 42, on an otherwise-empty RuleFiatAmount
    const cell = ruleCellFromBytes('RuleFiatAmount', data);
    expect(cell.kind).toBe('RawCell');
    expect(bytes(ruleCellToBytes('RuleFiatAmount', cell))).toEqual([0xc0, 0x0c, 0x2a]);
  });

  it('preserves a payload on a payload-free cell type verbatim', () => {
    const data = RuleFiatAmount.encode(
      RuleFiatAmount.fromPartial({ type: RuleFiatAmount_RuleFiatAmountType.RuleFiatAmountAny, payload: new Uint8Array([1, 2]) })
    ).finish();
    const cell = ruleCellFromBytes('RuleFiatAmount', data);
    expect(cell.kind).toBe('RawCell');
    expect(bytes(ruleCellToBytes('RuleFiatAmount', cell))).toEqual(bytes(data));
  });

  it('preserves an unknown cell-type enum across every family', () => {
    // Cell-type enum value newer than this SDK (cast to inject an out-of-range value).
    const cases: Array<[string, Uint8Array]> = [
      ['RuleFiatAmount', RuleFiatAmount.encode(RuleFiatAmount.fromPartial({ type: 902 as RuleFiatAmount_RuleFiatAmountType })).finish()],
      ['RuleIntegerGreater', RuleIntegerGreater.encode(RuleIntegerGreater.fromPartial({ type: 902 as RuleIntegerGreater_RuleIntegerGreaterType })).finish()],
      ['RuleUIntegerGreater', RuleUIntegerGreater.encode(RuleUIntegerGreater.fromPartial({ type: 902 as RuleUIntegerGreater_RuleUIntegerGreaterType })).finish()],
    ];
    for (const [col, data] of cases) {
      const cell = ruleCellFromBytes(col, data);
      expect(cell.kind).toBe('RawCell');
      expect(bytes(ruleCellToBytes(col, cell))).toEqual(bytes(data));
    }
  });

});

// --- Blockchain numeric passthrough ---

describe('rule cell codec blockchain numeric passthrough', () => {
  it('round-trips a contract-address cell with an unknown blockchain as a typed cell', () => {
    // ts-proto collapses an unknown blockchain enum to UNRECOGNIZED; the codec's
    // numeric passthrough keeps the raw number as a decimal string so the cell
    // stays typed and re-encodes byte-identically.
    const inner = RuleDestinationContractAddress.encode(
      RuleDestinationContractAddress.fromPartial({ address: '0x1', blockchain: 4242 as Blockchain })
    ).finish();
    const data = RuleDestination.encode(
      RuleDestination.fromPartial({ type: RuleDestination_RuleDestinationType.RuleDestinationContractAddress, payload: inner })
    ).finish();
    const cell = ruleCellFromBytes('RuleDestination', data);
    expect(cell).toEqual({ kind: 'DestinationContractAddress', address: '0x1', name: '', symbol: '', blockchain: '4242' });
    expect(bytes(ruleCellToBytes('RuleDestination', cell))).toEqual(bytes(data));
  });

  it('throws when encoding a contract-address cell with an unparseable blockchain name', () => {
    expect(() =>
      ruleCellToBytes('RuleDestination', { kind: 'DestinationContractAddress', address: '0x1', name: '', symbol: '', blockchain: 'NotAChain' })
    ).toThrow(/unknown blockchain/);
  });
});

// --- Magnitude encoding (via the public codec) ---

describe('rule cell codec magnitude encoding', () => {
  it('encodes the big-endian magnitude and round-trips zero', () => {
    const zero = ruleCellToBytes('RuleUIntegerGreater', { kind: 'UIntegerGreaterValue', value: 0n });
    expect(ruleCellFromBytes('RuleUIntegerGreater', zero)).toEqual({ kind: 'UIntegerGreaterValue', value: 0n });

    const big = RuleUIntegerGreater.decode(ruleCellToBytes('RuleUIntegerGreater', { kind: 'UIntegerGreaterValue', value: 256n }));
    expect(bytes(big.payload)).toEqual([0x01, 0x00]);

    // odd-length hex is left-padded to whole bytes
    const small = RuleIntegerGreater.decode(ruleCellToBytes('RuleIntegerGreater', { kind: 'IntegerGreaterValue', value: 8n }));
    expect(bytes(small.payload)).toEqual([0x08]);
  });

  it('selects the NegValue arm for negatives (sign carried by the arm, magnitude in the payload)', () => {
    const neg = RuleIntegerGreater.decode(ruleCellToBytes('RuleIntegerGreater', { kind: 'IntegerGreaterValue', value: -50n }));
    expect(neg.payload && bytes(neg.payload)).toEqual([50]);
    expect(ruleCellFromBytes('RuleIntegerGreater', ruleCellToBytes('RuleIntegerGreater', { kind: 'IntegerGreaterValue', value: -50n })))
      .toEqual({ kind: 'IntegerGreaterValue', value: -50n });
  });
});

// --- cellFamily (direct) ---

describe('cellFamily', () => {
  it('maps each cell kind to its column family', () => {
    const cases: Array<[RuleCell, string]> = [
      [{ kind: 'FiatAmountRange', minAmount: '1', maxAmount: '2' }, 'RuleFiatAmount'],
      [{ kind: 'SourceInternalWallet', path: 'p' }, 'RuleSource'],
      [{ kind: 'DestinationAny' }, 'RuleDestination'],
      [{ kind: 'StringEqualValue', value: 'v' }, 'RuleStringEqual'],
      [{ kind: 'BytesEqualValue', value: new Uint8Array() }, 'RuleBytesEqual'],
      [{ kind: 'StringArrayEqualValue', values: [] }, 'RuleStringArrayEqual'], // distinct from RuleStringEqual
      [{ kind: 'IntegerGreaterValue', value: 1n }, 'RuleIntegerGreater'],
      [{ kind: 'UIntegerGreaterValue', value: 1n }, 'RuleUIntegerGreater'], // distinct from RuleIntegerGreater
      [{ kind: 'WhitelistedContractAny' }, 'RuleWhitelistedContract'],
    ];
    for (const [cell, want] of cases) {
      expect(cellFamily(cell)).toBe(want);
    }
    // RawCell reports its own column type.
    expect(cellFamily({ kind: 'RawCell', columnType: 'RuleFiatAmount', payload: new Uint8Array() })).toBe('RuleFiatAmount');
  });
});

File: taurus-protect-sdk-java/client/src/test/java/com/taurushq/sdk/protect/client/mapper/RuleCellCodecTest.java
package com.taurushq.sdk.protect.client.mapper;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.math.BigInteger;
import java.util.Collections;

import com.google.protobuf.ByteString;
import com.google.protobuf.UnknownFieldSet;
import com.taurushq.sdk.protect.client.model.rulescontainer.RuleCell;
import com.taurushq.sdk.protect.proto.v1.RequestReply;
import org.junit.jupiter.api.Test;

/**
 * Branch and edge coverage for {@link RuleCellCodec}: encode error paths, the
 * RawCell lossless fallbacks, the magnitude helpers, {@code cellFamily}
 * resolution, and blockchain numeric passthrough.
 *
 * <p>Complements the shared golden-vector oracle ({@link GovernanceCellVectorsTest}),
 * which pins the happy-path wire bytes cross-SDK, and the container round-trip test.
 */
class RuleCellCodecTest {

    // --- Taxonomy 1: encode error branches ---

    @Test
    void encodeIntoWrongColumnFamilyThrows() {
        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
                () -> RuleCellCodec.encode("RuleFiatAmount", new RuleCell.StringEqualValue("x")));
        assertTrue(ex.getMessage().contains("is not valid for column type"));
    }

    @Test
    void encodeNegativeUnsignedIntegerThrows() {
        RuleCell[] negatives = {
                new RuleCell.UIntegerGreaterValue(BigInteger.valueOf(-1)),
                new RuleCell.UIntegerGreaterIsEqual(BigInteger.valueOf(-5)),
        };
        for (RuleCell cell : negatives) {
            IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
                    () -> RuleCellCodec.encode("RuleUIntegerGreater", cell));
            assertTrue(ex.getMessage().contains("requires a non-negative value"), cell.kind());
        }
    }

    @Test
    void encodeUnsupportedKindHitsTerminal() {
        // A recognised family prefix (FiatAmount) with no matching switch arm reaches
        // the encoder's terminal default; the empty columnType skips the family check.
        RuleCell bogus = new RuleCell() {
            @Override
            public String kind() {
                return "FiatAmountBogus";
            }
        };
        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
                () -> RuleCellCodec.encode("", bogus));
        assertTrue(ex.getMessage().contains("unsupported rule cell kind"));
    }

    @Test
    void cellFamilyUnknownKindThrows() {
        RuleCell zzz = new RuleCell() {
            @Override
            public String kind() {
                return "Zzz";
            }
        };
        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
                () -> RuleCellCodec.cellFamily(zzz));
        assertTrue(ex.getMessage().contains("unknown rule cell kind"));
    }

    // --- Taxonomy 2: RawCell fallback branches ---

    @Test
    void decodeUnknownColumnTypeReturnsRawVerbatim() {
        byte[] payload = {0x0a, 0x03, 'a', 'b', 'c'};
        RuleCell cell = RuleCellCodec.decode("SomeFutureColumn", payload);
        assertTrue(cell instanceof RuleCell.RawCell);
        assertEquals("SomeFutureColumn", ((RuleCell.RawCell) cell).columnType);
        assertArrayEquals(payload, RuleCellCodec.encode("SomeFutureColumn", cell));
    }

    @Test
    void decodeUnknownColumnTypeEmptyPayloadReturnsRaw() {
        // Java wraps even empty bytes of an unknown column in a RawCell (it does not
        // return null), so the cell survives a decode/encode cycle unchanged.
        RuleCell cell = RuleCellCodec.decode("SomeFutureColumn", new byte[0]);
        assertTrue(cell instanceof RuleCell.RawCell);
        assertArrayEquals(new byte[0], RuleCellCodec.encode("SomeFutureColumn", cell));
    }

    @Test
    void decodeKnownCellWithUnknownSubfieldFallsBackToRaw() {
        // A RuleFiatAmount carrying a protobuf field this SDK does not model cannot
        // round-trip through the typed FiatAmountRange, so the lossless guard keeps it
        // as a RawCell that re-encodes byte-for-byte.
        byte[] range = RequestReply.RuleFiatAmountRange.newBuilder()
                .setMinAmount("1000").setMaxAmount("50000").build().toByteArray();
        UnknownFieldSet uf = UnknownFieldSet.newBuilder()
                .addField(500, UnknownFieldSet.Field.newBuilder().addVarint(42).build())
                .build();
        byte[] withUnknown = RequestReply.RuleFiatAmount.newBuilder()
                .setType(RequestReply.RuleFiatAmount.RuleFiatAmountType.RuleFiatAmountRange)
                .setPayload(ByteString.copyFrom(range))
                .setUnknownFields(uf)
                .build().toByteArray();

        RuleCell cell = RuleCellCodec.decode("RuleFiatAmount", withUnknown);
        assertTrue(cell instanceof RuleCell.RawCell);
        assertArrayEquals(withUnknown, RuleCellCodec.encode("RuleFiatAmount", cell));
    }

    @Test
    void decodePayloadFreeCellCarryingPayloadFallsBackToRaw() {
        // A payload-free cell type (FiatAmountAny) that nonetheless carries a payload
        // does not round-trip through the typed value and is preserved as RawCell.
        byte[] anyWithPayload = RequestReply.RuleFiatAmount.newBuilder()
                .setType(RequestReply.RuleFiatAmount.RuleFiatAmountType.RuleFiatAmountAny)
                .setPayload(ByteString.copyFromUtf8("unexpected"))
                .build().toByteArray();

        RuleCell cell = RuleCellCodec.decode("RuleFiatAmount", anyWithPayload);
        assertTrue(cell instanceof RuleCell.RawCell);
        assertArrayEquals(anyWithPayload, RuleCellCodec.encode("RuleFiatAmount", cell));
    }

    // --- Taxonomy 3: magnitude helpers (package-private statics) ---

    @Test
    void magnitudeAndFromMagnitudeRoundTrip() {
        assertArrayEquals(new byte[] {0}, RuleCellCodec.magnitude(BigInteger.ZERO));
        // 128 == 0x80: BigInteger.toByteArray prepends a 0x00 sign byte; magnitude drops it.
        assertArrayEquals(new byte[] {(byte) 0x80}, RuleCellCodec.magnitude(BigInteger.valueOf(128)));
        // 256 == 0x0100: no spurious leading zero to drop.
        assertArrayEquals(new byte[] {0x01, 0x00}, RuleCellCodec.magnitude(BigInteger.valueOf(256)));

        assertEquals(BigInteger.ZERO, RuleCellCodec.fromMagnitude(new byte[] {0}));
        assertEquals(BigInteger.valueOf(128), RuleCellCodec.fromMagnitude(new byte[] {(byte) 0x80}));
        assertEquals(BigInteger.valueOf(256), RuleCellCodec.fromMagnitude(new byte[] {0x01, 0x00}));
    }

    @Test
    void unsignedIntegerZeroRoundTrips() {
        // The UIntegerGreater* zero path (only the signed-int zero is covered elsewhere).
        RuleCell zero = new RuleCell.UIntegerGreaterValue(BigInteger.ZERO);
        assertEquals(zero, RuleCellCodec.decode("RuleUIntegerGreater",
                RuleCellCodec.encode("RuleUIntegerGreater", zero)));
    }

    // --- Taxonomy 4: cellFamily resolution incl. the prefix-ordering guards ---

    @Test
    void cellFamilyResolvesEachColumn() {
        // StringArrayEqual* must resolve before StringEqual, and UIntegerGreater* before
        // IntegerGreater — pin the mapping for both pairs plus the RawCell echo.
        assertEquals("RuleStringArrayEqual",
                RuleCellCodec.cellFamily(new RuleCell.StringArrayEqualValue(Collections.singletonList("a"))));
        assertEquals("RuleStringEqual", RuleCellCodec.cellFamily(new RuleCell.StringEqualValue("a")));
        assertEquals("RuleUIntegerGreater", RuleCellCodec.cellFamily(new RuleCell.UIntegerGreaterValue(BigInteger.ONE)));
        assertEquals("RuleIntegerGreater", RuleCellCodec.cellFamily(new RuleCell.IntegerGreaterValue(BigInteger.ONE)));
        assertEquals("RuleFiatAmount", RuleCellCodec.cellFamily(new RuleCell.RawCell("RuleFiatAmount", new byte[0])));
    }

    // --- Taxonomy 5: blockchain numeric passthrough (cell-level) ---

    @Test
    void contractAddressBlockchainNumericPassthroughRoundTrips() {
        // A blockchain value unknown to this SDK survives as its numeric string form.
        RuleCell.DestinationContractAddress original =
                new RuleCell.DestinationContractAddress("0xabc", "Token", "TKN", "4242");
        RuleCell decoded = RuleCellCodec.decode("RuleDestination",
                RuleCellCodec.encode("RuleDestination", original));
        assertEquals(original, decoded);
        assertEquals("4242", ((RuleCell.DestinationContractAddress) decoded).blockchain);
    }

    @Test
    void contractAddressUnparseableBlockchainThrows() {
        // A non-numeric, non-enum blockchain name is a caller error surfaced by parseInt.
        assertThrows(NumberFormatException.class, () -> RuleCellCodec.encode("RuleDestination",
                new RuleCell.DestinationContractAddress("0xabc", "Token", "TKN", "NotAChain")));
    }
}
